src/liballoc/collections/vec_deque.rs

```diff
@@ -62,6 +62,64 @@ impl<T: Clone> Clone for VecDeque<T> {
     fn clone(&self) -> VecDeque<T> {
         self.iter().cloned().collect()
     }
+
+    fn clone_from(&mut self, other: &Self) {
+        self.truncate(other.len());
+
+        let len = self.len();
+        let other_len = other.len();
+        if len < other_len {
+            self.reserve_exact(other_len - len);
+        }
+
+        let contiguous = self.is_contiguous();
+        let other_contiguous = other.is_contiguous();
+
+        let (s_front, s_back) = self.as_mut_slices();
+        let (o_front, o_back) = other.as_slices();
+
+        let s_front_len = s_front.len();
+        let o_front_len = o_front.len();
+        // len <= other_len, so only use len elements from other
+        let only_using_o_front = other_contiguous || len <= o_front_len;
+        let cap = len.saturating_sub(o_front_len);
+        match (contiguous, only_using_o_front, s_front_len.cmp(&o_front_len)) {
+            (true, true, _) => {
+                // dst contiguous, src contiguous
+                s_front.clone_from_slice(&o_front[..len]);
+            },
+            (true, false, _) => {
+                // dst contiguous, src wraps
+                s_front[..o_front_len].clone_from_slice(&o_front);
+                s_front[o_front_len..].clone_from_slice(&o_back[..cap]);
+            },
+            (false, true, _) => {
+                // dst wraps, src contiguous
+                s_front.clone_from_slice(&o_front[..s_front_len]);
+                s_back.clone_from_slice(&o_front[s_front_len..len]);
+            },
+            (false, false, Ordering::Equal) => {
+                // dst wraps, src wraps, both fronts same length
+                s_front.clone_from_slice(&o_front);
+                s_back.clone_from_slice(&o_back[..cap]);
+            },
+            (false, false, Ordering::Greater) => {
+                // dst wraps, src wraps, s_front longer than o_front
+                let overlap = s_front_len - o_front_len;
+                s_front[..o_front_len].clone_from_slice(&o_front);
+                s_front[o_front_len..].clone_from_slice(&o_back[..overlap]);
+                s_back.clone_from_slice(&o_back[overlap..cap]);
+            },
+            (false, false, Ordering::Less) => {
+                // dst wraps, src wraps, o_front longer than s_front
+                let overlap = o_front_len - s_front_len;
+                s_front.clone_from_slice(&o_front[..s_front_len]);
+                s_back[..overlap].clone_from_slice(&o_front[s_front_len..]);
+                s_back[overlap..].clone_from_slice(&o_back[..cap]);
+            },
```

Comment:
I was wondering if we have to stem the bloat of branches here, to have just 1 branch, not 6.

Sorry for writing so much code, I ended up prototyping that like this - https://play.rust-lang.org/?version=stable&mode=debug&edition=2018&gist=6ff3024edcb82d8f3c605963f0ba9dbf

It's not necessarily a better approach. I'd like to test compile both and look at the compile time/code size. The code in the PR would have a huge code size if the `clone_from_slice` calls were inlined, so that's the main thing I'd be afraid of.

Reply:
I really like that iterator. There is a private helper struct for iteration in btree as well, so it's not particularly bizarre (I just checked because I hadn't considered it and it's very satisfying). I'm not sure about the compile time/code size differences, but I do think the whole approach is better if it's comparable in that respect.

Bounds checks from the clone_from_slice? Because those should be the same either way, I think. Otherwise I'm not sure.

The only other idea I have is to go with something closer to the original. I know it doesn't really take advantage of a lot of optimizations, but if both the matching approach and the iterator end up being considered more complicated than they're worth I think it would still be good to have access to clone_from implementations for the contained type.

Separately, the test in your main is much more concise than my test; it takes longer but I'm not sure how much that matters. Should I change my test to a similar nested loop (maybe just to 6 or 8 rather than 16)?

Reply:
Yes “exhaustive” style tests are good, I haven't really checked that this test covers all cases, but to I think it's a good idea to test using that method.

W.r.t bounds checks, they are functionally equivalent but have different context in the code, there are fewer locations where they happen, where we are splitting in PairSlices, and once to check equal lengths in `clone_from_slice`. (With luck they could even compile out, based on the `min` calculation?).

Reply:
Mostly I mention the original because in my (admittedly casual) benchmarks clone_from can be faster than clone by 3-10x, but the two clone_from implementations (I haven't tested the custom iterator version yet) are within each other's margins of error.

Reply:
Yes, we can switch!

I've compared the two new impls for generated code, and the code in my playground link (+ reserve call which I added) seems to compile to simpler smaller code.

As a comparison, A has 13 calls to memcpy inside release compile of VecDeque::<i32>::clone_from and B has 3 calls to memcpy.  A has 14 panic sites and B has 3 etc.

I haven't benchmarked or looked at compile time.

Reply:
I changed everything over, tests work and benchmarks are at least as good.

Clone still outperforms clone_from when dst is much larger than src and T is copy, though; is there an argument against using the same optimization as vec's truncate for types that don't need to be dropped? If not, I would like to open a pull request for adding it after this one lands (since it might need testing but is a fairly small change). Something like this:
```rust
pub fn truncate(&mut self, len: usize) {
    let current_len = self.len();
    if mem::needs_drop::<T>() {
        for _ in len..current_len {
            self.pop_back();
        }
    } else if len < current_len {
        self.head = self.wrap_sub(self.head, current_len - len);
    }
}
```
This change makes clone_from at least as fast as clone in all cases, and it just seems reasonable to take advantage of contiguous memory this way (unless there's some cost I'm not seeing)

Reply:
Using either `needs_drop` or `drop_in_place(*mut [T])` would be good in truncate, preferably the latter if possible. That's what Vec::truncate wants to change to using too, if possible.

Reply:
I think `drop_in_place` would make sense and wouldn't be too hard. It seems like #64432 is going to get merged, and while VecDeque can't use exactly the same argument (`clear` uses `drain` for a VecDeque instead of `truncate(0)`), I think it makes sense to fix truncate and start using it in `clear` as well